### DISCUSSION TITLE
2D egui on top of 3D Rend3, etc. - who has the mouse?

### DISCUSSION THREAD

**Discussion by John-Nagle** [emilk/egui#1718] [Q&A]
I'm using egui, rend3-egui, and egui_winit_platform to put a 2D menu-type interface on top of a 3D window. Works fine.
I'm working out some of the issues over how the two systems share the mouse. 
It seems that egui just polls the mouse on each frame; it doesn't receive mouse window events. 3D land gets window events and mouse position too, and I need to know if egui is using the mouse. If it is, mouse window events should be ignored by the 3D interface.

![eguitest02](https://user-images.githubusercontent.com/942608/172030871-fc7d62b1-533c-4506-a391-c39a28a095c0.png)

**File menu has dropped, and cursor is not over it.** I need to know that a menu is active. [egui](https://docs.rs/egui/latest/egui/index.html)::[output](https://docs.rs/egui/latest/egui/output/index.html)::[OutputEvent](https://docs.rs/egui/latest/egui/output/enum.OutputEvent.html#) almost does what I want, but there's no "menu is dropped" output event. (That's built from the Rend3 "egui" demo, in case you want to run something like this.)

My goal is a clean game screen. I want to do something like what YouTube does in the player user interface. When the user isn't using the 2D UI, it should disappear. Moving the mouse to the top or bottom will bring it back. That's part of why I need to know if the 2D UI is in use. "In use" can be be defined as the cursor being over any egui element, or any dynamically displayed egui element being visible. I don't want to disappear the egui display while someone is navigating a submenu or something.

I'd also like, when a window is full screen, to get rid of the header bar. There's a call to do that at the "winit" level, but should I be calling that when using these other crates on top?

On an unrelated note, the Rust icon is a .png with an alpha channel transparency, and should be transparent where it is white. Not sure what I'm doing wrong there, but the "egui" example for Rend3 does the same thing.

**Comment by John-Nagle** [comment]
I've solved part of the problem, but not all of it. I can tell if the cursor is over a menu item by checking "hovering". But I have to write code to do this for every menu item. Like this:

```
        if egui::TopBottomPanel::top("menu_bar").show(&ctx, |ui| {
            menu::bar(ui, |ui| {
                ui.menu_button("File", |ui| {
                    let response = ui.button("Open");
                    if response.clicked() {
                        if let Some(path) = rfd::FileDialog::new()
                            .set_title("Viewer session file to play back")
                            .add_filter("json", &["json"])
                            .pick_file()
                        {
                            let picked_path = Some(path.display().to_string());
                            println!("File picked: {}", picked_path.unwrap());
                        }
                    }
                    if response.hovered() { inuse = true; }     // pointer is over this button
                });
                ui.menu_button("Help", |ui| {
                    let response = ui.button("Help");
             	    if response.clicked() {
                        webbrowser::open("https://www.rust-lang.org").expect("failed to open URL");
                    }
                    if response.hovered() { inuse = true; }
                });
            });
        }).response.hovered() { inuse = true };

```
Works, but it gets verbose. I'd like to have a function that tells me if an object **or any of its children** are being hovered over.

**Comment by emilk** [comment]
Sounds like you are looking for https://docs.rs/egui/latest/egui/struct.Context.html#method.is_using_pointer and https://docs.rs/egui/latest/egui/struct.Context.html#method.is_pointer_over_area ?

**Reply by John-Nagle** [reply]
[is_pointer_over_area](https://docs.rs/egui/latest/egui/struct.Context.html#method.is_pointer_over_area) - that's the convenience function I needed. Thanks.
